source/simulator/solver_schemes.cc

```diff
@@ -0,0 +1,744 @@
+
+        pcout << "      Relative Stokes residual after nonlinear iteration " << nonlinear_iteration+1
+              << ": " << relative_stokes_residual
+              << std::endl;
+
+        if (parameters.run_postprocessors_on_nonlinear_iterations)
+          postprocess ();
+
+        // if reached convergence, exit nonlinear iterations.
+        if (relative_stokes_residual < parameters.nonlinear_tolerance)
+          break;
+
+        pcout << std::endl;
+
+        ++nonlinear_iteration;
+      }
+    while (nonlinear_iteration < max_nonlinear_iterations);
+
+    return;
+  }
+
+
+  template <int dim>
+  void Simulator<dim>::solve_Newton_Stokes ()
+  {
+    std::vector<double> initial_composition_residual (parameters.n_compositional_fields,0);
+
+    double initial_residual = 1;
+
+    double velocity_residual = 0;
+    double pressure_residual = 0;
+    double residual = 1;
+    double residual_old = 1;
+
+    double switch_initial_residual = 1;
+    double newton_residual_for_derivative_scaling_factor = 1;
+
+    bool use_picard = true;
+
+    const unsigned int max_nonlinear_iterations =
+      (pre_refinement_step < parameters.initial_adaptive_refinement)
+      ?
+      std::min(parameters.max_nonlinear_iterations,
+               parameters.max_nonlinear_iterations_in_prerefinement)
+      :
+      parameters.max_nonlinear_iterations;
+
+    double stokes_residual = 0;
+    for (nonlinear_iteration = 0; nonlinear_iteration < max_nonlinear_iterations; ++nonlinear_iteration)
+      {
+        assemble_and_solve_temperature();
+        assemble_and_solve_composition();
+
+        assemble_newton_stokes_system = true;
+
+        solution.block(introspection.block_indices.pressure) = 0;
+        solution.block(introspection.block_indices.velocities) = 0;
+
+        if (use_picard == true && (residual/initial_residual <= parameters.nonlinear_switch_tolerance ||
+                                   nonlinear_iteration >= parameters.max_pre_newton_nonlinear_iterations))
+          {
+            use_picard = false;
+            pcout << "   Switching from defect correction form of Picard to the Newton solver scheme." << std::endl;
+
+            /**
+             * This method allows to slowly introduce the derivatives based
+             * on the improvement of the residual. If we do not use it, we
+             * just set it so the newton_derivative_scaling_factor goes from
+             * zero to one when switching on the Newton solver.
+             */
+            if (!parameters.use_newton_residual_scaling_method)
+              newton_residual_for_derivative_scaling_factor = 0;
+          }
+
+        newton_handler->set_newton_derivative_scaling_factor(std::max(0.0,
+                                                                      (1.0-(newton_residual_for_derivative_scaling_factor/switch_initial_residual))));
+
+
+        /**
+         * copied from solver.cc
+         */
```

Comment:
is this comment still accurate?

Reply:
I do not know, I just copied that. I think @MFraters introduced it. It looks like a (partly modified) copy of `solve_stokes` though.

Reply:
I think it is still quite accurate, but not sure whether it is necessary to state it like this here. Just for a background: it had to be copied (and slightly modified) because of the denormalisation.